Add Champions Unofficial compat

Hypocisy opened the pull request:
## Add champions Compat
- Add Champions Compat, if entity aready champion mob, stop process spawning entity to inf mob
- We should check if entity is champion, not only check champion entity's persistent data, because when champion created, entity does not have persistent data yet.
- in this case, I made champion spawn event to HIGHEST EventPriority，so forge should first let entity become champion, and then the inf mod will check if champion entity persistet has PersistentData, but no persistentData to check.
- Please see those pictures
- ![image](https://github.com/user-attachments/assets/47c91d9b-24ee-42e4-8c88-06e3f1c84315)
- ![image](https://github.com/user-attachments/assets/59e9427f-bd88-4b22-838d-ae1bee68d20d)
- ![image](https://github.com/user-attachments/assets/abdd1336-4e11-4d20-b5a7-a4a9606d5c7a)

Hypocisy committed:
- We should check if entity is champion, not only check champion entity's persistent data, because when champion created, entity does not have persistent data yet.
- Only do spawn process and entity attributes modifier if entity joined world
- Add ChampionsCompat
- use entity random source instead level random
- Upgrade project forge version
- Simply the logger usage

Hypocisy committed:
- fix

AtomicStryker commented:
First, thank you for participating

The "code cleaning" has made the PR much bigger and harder to read, sadly.
It would have maybe been better to split this into 2 PR, one for "functionality" and one for "code style changes"

Generally speaking, i am very hesitant to add hard dependencies to my mods because in the past, dependencies always stopped updating or updated later etc

Hypocisy commented:
it's not hard dependencies, it's soft at all.
with out champions, your mod still can run property

Hypocisy commented:
sorry for that, i will split it to 2 PR

Hypocisy committed:
- Add champions compat

Hypocisy committed:
- Remove unnecessary changes

Hypocisy commented:
@AtomicStryker Well, right now this pr only contain champion compat related code

AtomicStryker commented on InfernalMobs/src/main/java/atomicstryker/infernalmobs/common/InfernalMobsCore.java:
```diff
@@ -114,6 +80,7 @@ public InfernalMobsCore() {
         classesForcedMap = new HashMap<>();
         classesHealthMap = new HashMap<>();
         modifiedPlayerTimes = new HashMap<>();
+        championLoaded = ModList.get().isLoaded("champions");
```
you can probably pull the modID from the actual ChampionsCore class, rather than hardcoding a magic string?

Hypocisy replied:
Yes, i will change it to non magic string

AtomicStryker commented on InfernalMobs/src/main/java/atomicstryker/infernalmobs/common/EntityEventHandler.java:
```diff
@@ -31,16 +32,28 @@ public EntityEventHandler() {
+        // check level is server level first
+        if (!level.isClientSide()) {
+            // champions mob compat start
+            if (InfernalMobsCore.instance().isChampionLoaded() && ChampionsCompat.isChampionEntity(entity)) {
```
should add a boolean configurable that ALLOWS a mob to be both champion and infernal. masochists will ask for it

Hypocisy replied:
Alright, let’s add a config to let user choose enable or not enable compatibility, add config let user choose the entity becomes inf or champions, or add chance to let game choose it self.

AtomicStryker commented on InfernalMobs/build.gradle:
```diff
@@ -74,8 +94,33 @@ minecraft {
+    annotationProcessor "org.spongepowered:mixin:${mixin_version}:processor"
+
+    compileOnly fg.deobf("curse.maven:champions-unofficial-1074990:${cf_champions}")
+    compileOnly fg.deobf("curse.maven:jade-324717:${cf_jade}")
```
whats jade? some kind of HUD? why depend on it? you dont appear to be importing any classes from it anywhere

Hypocisy replied:
It’s a soft depend with champions I just used it for test, but it does not need in this pr, will fixed later

AtomicStryker commented on InfernalMobs/gradle.properties:
```diff
@@ -58,4 +58,14 @@ mod_authors=AtomicStryker
+curseforge_id=227875
+
+# mod depends
+cf_champions=6133394
```
this is a specific version of champions, yes? please write how to get this number from curseforge files so this is maintainable

Hypocisy replied:
Yes it’s specific version of champions, we are using the curse maven, this maven is providing by curse forge, you can find any of the curseforge mod jar to import like champions unofficial: https://www.curseforge.com/minecraft/mc-mods/champions-unofficial/files/6136541
Here to copy the full line: 
![IMG_6279](https://github.com/user-attachments/assets/7f019a25-ea98-4139-bb45-0114a07507dc)
implementation fg.deobf("curse.maven:champions-unofficial-1074990:6133394")

so 1074990 is cf project id, 6133394 is file id, the project id is hard coded since we don’t really change this, the file id is stored by the variable cf_champions

But we just wanted this magic number 6133394

so paste it to
cf_champions=

you can add more depends like this way for better and easy maintenance.
And rebuild or reimport project, then champions(any project) will download automatically.

AtomicStryker commented on InfernalMobs/build.gradle:
```diff
@@ -74,8 +94,33 @@ minecraft {
 dependencies {
     minecraft "net.minecraftforge:forge:${minecraft_version}-${forge_version}"
+
+    compileOnly(annotationProcessor("io.github.llamalad7:mixinextras-common:${mixin_extras_version}"))
```
is all of this really needed? we are not compiling champions, we just want its jar and read one class from it, but i really dont have much experience with sponge/mixins

Hypocisy replied:
This line is not needed, I will make a change to fix this.
and yes, if you want other mod that contain mixin to run in dev for test, you need to add those lines to avoid mixin  inject error, that will stop you  build own mod

AtomicStryker reviewed (changes requested):
already much cleaner looking! i think there is still more in your PR than is required

Hypocisy committed:
- Removed unnecessary changes

Hypocisy committed:
- Updated the Champions mod to use the new API, allowing it to track whether an entity is a champion and independently manage affixes, rank, and entity modifiers.
- Added `enableChampionCompat` (boolean) to toggle compatibility with the Champions mod.
- Added `useInfernalMobs` (boolean) to allow or prevent Infernal Mobs from handling champion entities.
- Added `useRandomChampionSelection` (boolean) to enable probability-based entity handling.
- Added `championSelectionChance` (double, 0-100) to configure the probability of Champions handling entities instead of Infernal Mobs.
- Refactored `shouldChampionHandleEntity()` to integrate the new configuration options.

Hypocisy committed:
- fix

Hypocisy committed:
Update InfernalMobsCore.java

Rename

Hypocisy committed:
- rename

AtomicStryker commented on InfernalMobs/src/main/java/atomicstryker/infernalmobs/common/InfernalMobsConfig.java:
```diff
@@ -185,4 +189,36 @@ public List<String> getEntityTagBlackList() {
     }
+
+    public boolean getUseInfernalMobs() {
+        return useInfernalMobs;
```
this name does not describe what the option does very well. "championMobCanBecomeInfernal" would be better, for example

Hypocisy replied:
I wanted to create a separate config file specifically for mod compatibility settings. However, after completing it, I found that the config file was never generated in the expected location, so I gave up and reverted to using the original file. But I forgot to rename it at that time. Later, when I did rename it, I didn’t understand why it was still not recognized correctly.

AtomicStryker replied:
once a runtime has generated a config file, you will have to delete it when making code changes. otherwise the gson deserializer will probably stick with defaults and ignore unknowns

EDIT: For the record, i dont use forge/mc provided config files because the stupid things change every 2 versions and i am sick of refactoring. this gson thing does not ever change until i change it

Hypocisy replied:
I tried to delete, but only gen the default one, i use the same way as defaut config, but never work, weird. let's just use the default config lol

AtomicStryker commented on InfernalMobs/src/main/resources/changelog.txt:
```diff
@@ -1,3 +1,7 @@
+1.20.1.5
+- Add Champions Compat, if entity already champion mob, stop process spawning entity to inf mob
```
best mention each new config value explicitly in the changelog

Hypocisy replied:
i forgot, i will write now

Hypocisy replied:
i writed, please check it

AtomicStryker reviewed (changes requested):
already looking much better

Hypocisy committed:
- Add allow both Champions and InfernalMobs Config options

Hypocisy committed:
- update changelog.txt

Hypocisy committed:
- update changelog.txt
- rename useInfernalMobs to championMobCanBecomeInfernal

Hypocisy committed:
Merge remote-tracking branch 'github/1.20.1' into 1.20.1

# Conflicts:
#	InfernalMobs/src/main/resources/changelog.txt

AtomicStryker commented on InfernalMobs/src/main/resources/changelog.txt:
```diff
@@ -1,3 +1,15 @@
+- Added flexible config allow user control mob will handel by champions or Infernal mob, or handel by both
+- Updated the Champions mod to use the new API, allowing it to track whether an entity is a champion and independently manage affixes, rank, and entity modifiers.
+- Added `enableChampionCompat` (boolean) to toggle compatibility with the Champions mod.(default true)
+- Added `useInfernalMobs` (boolean) to allow or prevent Infernal Mobs from handling champion entities.(default true)
```
please rename this to "infernalMobsReplacesChampions" as this more correctly describes what it does

Hypocisy replied:
okay

AtomicStryker commented on InfernalMobs/src/main/resources/changelog.txt:
```diff
@@ -1,3 +1,15 @@
+1.20.1.5
+- Added Champions Compat
+- Added flexible config allow user control mob will handel by champions or Infernal mob, or handel by both
```
will be handled, handled by both

Hypocisy committed:
- more rename and Correction

Hypocisy committed:
- more rename and Correction

AtomicStryker commented on InfernalMobs/src/main/resources/changelog.txt:
```diff
@@ -1,3 +1,15 @@
+- Added `infernalMobsReplacesChampions` (boolean) to allow or prevent Infernal Mobs from handling champion entities.(default true)
+- Added `useRandomChampionSelection` (boolean) to enable probability-based entity handling.(default false)
+- Added `championSelectionChance` (double, 0-100) to configure the probability of Champions handling entities instead of Infernal Mobs.(default 50.0)
+- Added `allowBothChampionAndInfernal` to configure mob handel by both mod(default false), for masochists.(default false)
```
"mob being changed by both mods", and you write default twice

Hypocisy replied:
sorry, my english is not good

AtomicStryker replied:
it is perfectly sufficient, and i will point out grammar issues i want fixed ;)

Hypocisy committed:
Merge remote-tracking branch 'github/1.20.1' into 1.20.1

AtomicStryker commented:
i checked out the merged version, built it, tested it locally, looked good to me. great job!

AtomicStryker commented:
and 1.20.1.5 is pushed to curse. it will take a little while and then show up there

Hypocisy commented:
> i checked out the merged version, built it, tested it locally, looked good to me. great job!

thanks for your time, you are so kind and good guy

Hypocisy commented:
> and 1.20.1.5 is pushed to curse. it will take a little while and then show up there

yes, Cheers!!!
